src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,705 @@
+            R current = initial.get();
+            boolean integrate(T element, Downstream<? super R> downstream) {
+                return downstream.push(current = scanner.apply(current, element));
+            }
+        }
+
+        return Gatherer.ofSequential(State::new,
+                Integrator.<State,T, R>ofGreedy(State::integrate));
+    }
+
+    /**
+     * An operation which executes a function concurrently
+     * with a configured level of max concurrency, using
+     * <a href="{@docRoot}/java.base/java/lang/Thread.html#virtual-threads">virtual threads</a>.
+     * This operation preserves the ordering of the stream.
+     *
+     * @apiNote In progress tasks will be attempted to be cancelled,
+     * on a best-effort basis, in situations where the downstream no longer
+     * wants to receive any more elements.
+     *
+     * @implSpec If a result of the function is to be pushed downstream but
+     * instead the function completed exceptionally then the corresponding
+     * exception will instead be rethrown by this method as an instance of
+     * {@link RuntimeException}, after which any remaining tasks are canceled.
+     *
+     * @param maxConcurrency the maximum concurrency desired
+     * @param mapper a function to be executed concurrently
+     * @param <T> the type of input
+     * @param <R> the type of output
+     * @return a new Gatherer
+     * @throws IllegalArgumentException if {@code maxConcurrency} is less than 1
+     * @throws NullPointerException if {@code mapper} is {@code null}
+     */
+    public static <T, R> Gatherer<T,?,R> mapConcurrent(
+            final int maxConcurrency,
+            final Function<? super T, ? extends R> mapper) {
+        if (maxConcurrency < 1)
+            throw new IllegalArgumentException(
+                    "'maxConcurrency' must be greater than 0");
+
+        Objects.requireNonNull(mapper, "'mapper' must not be null");
+
+        class State {
+            final ArrayDeque<Future<R>> window = new ArrayDeque<>(maxConcurrency);
+            final Semaphore windowLock = new Semaphore(maxConcurrency);
+
+            final boolean integrate(T element,
+                                    Downstream<? super R> downstream) {
+                if (!downstream.isRejecting())
+                    createTaskFor(element);
+                return flush(0, downstream);
+            }
+
+            final void createTaskFor(T element) {
+                windowLock.acquireUninterruptibly();
+
+                var task = new FutureTask<R>(() -> {
+                    try {
+                        return mapper.apply(element);
+                    } finally {
+                        windowLock.release();
+                    }
+                });
+
+                var wasAddedToWindow = window.add(task);
+                assert wasAddedToWindow;
+
+                Thread.startVirtualThread(task);
+            }
+
+            final boolean flush(long atLeastN,
```

Comment:
@AlanBateman @PaulSandoz Was alerted to odd nesting of exceptions so I reworked the exception handling of mapConcurrent to unpack any ExecutionException, and intercept InterruptedExceptions to cancel remaining work and reinstating the interrupted flag.